cpu/stm32/periph/spi.c

```diff
@@ -162,7 +178,7 @@ int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
     /* enable SPI device clock */
     periph_clk_en(spi_config[bus].apbbus, spi_config[bus].rccmask);
     /* enable device */
-    uint8_t br = spi_divtable[spi_config[bus].apbbus][clk];
+    uint8_t br = _get_clkdiv(&spi_config[bus], clk);
```

Comment:
I would prefer to do that in the spi_init function otherwise this add several instructions each time a driver tries to acquire the SPI bus.

I also tried to do that some days ago in https://github.com/aabadie/RIOT/commit/879d7172727f35e34857366b07eae8870ec4d85f. The logic is a bit different: it reuses the same algorithm used by the spi_divtable tool and is optional (because it adds 120B on ROM).
Maybe we could use the algorithm in `_get_clkdiv` that I find simpler than the original one. 
Since the SPI clock speed is highly dependent on the bus clock, if one wants to reach a precise spi clock speed, we should provide more control on the core clock and peripheral bus dividers parameters (using CFLAGS and Kconfig), in parallel with runtime computation of the BR (I wanted to answer that in #14731 initially because I think it's the best approach).

Reply:
> I would prefer to do that in the spi_init function otherwise this add several instructions each time a driver tries to acquire the SPI bus.

Yeah, that's also my main concern with this.

With this PR I'm measuring 5.97 microsecond for acquire/release on 10 MHz, 7.32 microsecond on 100 kHz. On master I'm measuring 5.39 microsecond on 10 MHz and 100 kHz. I think I should be able to speed it up a bit, but not by much.

> I also tried to do that some days ago in aabadie@879d717. The logic is a bit different: it reuses the same algorithm used by the spi_divtable tool and is optional (because it adds 120B on ROM).
Maybe we could use the algorithm in _get_clkdiv that I find simpler than the original one.

I think we should be able to manage with just 7 `uint32_t` values per bus, indicating the clock speed boundaries between the divider values. So 56B for both APB1 and APB2. This would still require iterating over the array, so I'm not sure if it is actually faster than the fully dynamic case.

> Since the SPI clock speed is highly dependent on the bus clock, if one wants to reach a precise spi clock speed, we should provide more control on the core clock and peripheral bus dividers parameters (using CFLAGS and Kconfig), in parallel with runtime computation of the BR

I agree, and I consider it out-of-bounds here. My main goal here is to remove the need for the precomputed tables and allow for arbitrary clock speeds.

Reply:
> My main goal here is to remove the need for the precomputed tables and allow for arbitrary clock speeds.

I also initially wanted to remove the need for precomputed tables because I find them too static. But they have an advantage in term of code size and are also a bit more performant (no iteration, just an access to a value in a table) so I thought both could coexist: the precomputed tables give general purpose defaults and dynamic computation + configurable clock speeds gives more control for people with more specific needs. This is why in https://github.com/aabadie/RIOT/commit/879d717 I introduced the USE_CUSTOM_CLOCK define, which is disabled by default.
I also agree that the configurable clock speed is not needed for such a PR and could be implemented later (and requires different work for the different families).

> that's also my main concern with this

I also forgot to mention that for each acquire, you are recomputing the same values, having them computed only once during the init phase seems more efficient to me.

Reply:
How about we store the last clock speed and only compute a new value if it it is different from the last call?

Reply:
> How about we store the last clock speed and only compute a new value if it it is different from the last call?
> Then we can also skip writing the hardware registers each time.

I like this idea a lot. What do you think @aabadie?

Reply:
I think that this would work well when there's only one SPI device connected or several devices and all using the same SPI clock speed. So that would introduce inconsistent behavior depending on the setup (worst case when several devices are using different clock speed). I don't know how bad or good this is.

Reply:
The common case is just one device / bus.
If there are many devices *and* they use different clock speeds, the only penalty would be an increase in `aquire` time.

Right now we incur that penalty on every user.

Reply:
@bergzand it would be nice if you could try the clock speed caching strategy and give some numbers. If it's clearly better than in master, then I think we should go for it.
One minor drawback I see is that there's no way to know which is the real spi clock speed (with the divtable this is know by reading the board configuration). Maybe printing it in a debug statement in `_spi_clkdiv` could be used as a workaround.

Reply:
> @bergzand it would be nice if you could try the clock speed caching strategy and give some numbers. If it's clearly better than in master, then I think we should go for it.

Unfortunately it is more complex than this. The clock divider register is shared with the SPI mode and SPI chip select configuration settings. So only if we're sure that not only the clock speed matches the previous configuration, but also the mode and whether soft CS is used we can skip the register write.

Caching the clock divider can still save approx 450ns (see below) per acquire/release.

Reply:
With caching the average acquire/release timing for 1000 acquire/releases is 5.54 microsecond.